File: lab4/task1/cpp/main.cpp
#include <iostream>
#include <vector>
using namespace std;

class Product
{
private:
    string name;
    double price_rub;
    string manufacturer;

public:
    Product() : name(""), price_rub(0.0), manufacturer("") {}

    Product(string n, double p, string m) : name(move(n)), price_rub(p), manufacturer(move(m)) {}

    void input()
    {
        cout << "Введите наименование (0 - выход): ";
        getline(cin, name);
        if (name.empty())
            name = "0";
        if (name == "0")
            return;

        cout << "Введите цену (в рублях): ";
        string input_str;
        getline(cin, input_str);
        if (input_str.empty())
            input_str = "0";

        try
        {
            price_rub = stod(input_str);
        }
        catch (...)
        {
            price_rub = 0;
        }

        cout << "Введите изготовителя: ";
        getline(cin, manufacturer);
        if (manufacturer.empty())
            manufacturer = "Неизвестный изготовитель";
    }

    void output() const
    {
        cout << "Товар: " << name
             << ", Цена: " << price_rub
             << " руб., Изготовитель: " << manufacturer << endl;
    }

    double to_usd(double rate = 100.0) const
    {
        return price_rub / rate;
    }

    void increase_if_toyota(double rate = 100.0, double increase = 50.0)
    {
        if (to_lower_case(name).find("toyota") != string::npos)
            price_rub += increase * rate;
    }

    bool is_exit() const
    {
        return name == "0";
    }

    static string to_lower_case(string str)
    {
        for (char &c : str)
            c = static_cast<char>(tolower(static_cast<unsigned char>(c)));
        return str;
    }
};

int main()
{
    vector<Product> product_list;

    while (true)
    {
        Product product;
        product.input();
        if (product.is_exit())
            break;
        product_list.push_back(product);
    }

    cout << "\n=== Список товаров ===\n";
    for (auto &product : product_list)
    {
        product.output();
        cout << "Цена в долларах: " << product.to_usd() << endl;
        product.increase_if_toyota();
        cout << "После проверки: ";
        product.output();
        cout << endl;
    }

    return 0;
}

File: lab4/task1/java/Main.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        ArrayList<Product> productList = new ArrayList<>();

        while (true) {
            Product product = new Product();
            product.input(scanner);
            if (product.isExit())
                break;
            productList.add(product);
        }

        System.out.println("\n=== Список товаров ===");
        for (Product product : productList) {
            product.output();
            System.out.println("Цена в долларах: " + product.toUSD(100));
            product.increaseIfToyota(100, 50);
            System.out.print("После проверки: ");
            product.output();
            System.out.println();
        }
    }
}

class Product {
    private String name;
    private double priceRub;
    private String manufacturer;

    public Product() {
    }

    public Product(String name, double priceRub, String manufacturer) {
        this.name = name;
        this.priceRub = priceRub;
        this.manufacturer = manufacturer;
    }

    public void input(Scanner scanner) {
        System.out.print("Введите наименование (0 - выход): ");
        name = scanner.nextLine();
        if (name.isEmpty())
            name = "0";
        if (name.equals("0"))
            return;

        System.out.print("Введите цену (в рублях): ");
        String priceStr = scanner.nextLine();
        if (priceStr.isEmpty())
            priceStr = "0";
        try {
            priceRub = Double.parseDouble(priceStr);
        } catch (NumberFormatException e) {
            priceRub = 0;
        }

        System.out.print("Введите изготовителя: ");
        manufacturer = scanner.nextLine();
        if (manufacturer.isEmpty())
            manufacturer = "Неизвестный изготовитель";
    }

    public void output() {
        System.out.println("Товар: " + name + ", Цена: " + priceRub + " руб., Изготовитель: " + manufacturer);
    }

    public double toUSD(double rate) {
        return priceRub / rate;
    }

    public void increaseIfToyota(double rate, double increase) {
        if (name.toLowerCase(java.util.Locale.ROOT).contains("toyota"))
            priceRub += increase * rate;
    }

    public boolean isExit() {
        return name.equals("0");
    }
}
